File: offer/JZ28 对称的二叉树/tree.java
/*
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    boolean isMatching(TreeNode a, TreeNode b){
        if(a == null && b == null)
            return true;
        if(a == null && b != null)
            return false;
        if(a != null && b == null)
            return false;
        if(a.val != b.val)
            return false;
        return isMatching(a.left, b.right) && isMatching(a.right, b.left);
    }
    boolean isSymmetrical(TreeNode root) {
        if(root == null)
            return true;
        return isMatching(root.left, root.right);
    }
}

File: 0538. Convert BST to Greater Tree/tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* convertBST(TreeNode* root) {
        dfs(root, 0);
        return root;
    }
    
    int dfs(TreeNode* root, int fromParent){
        if(root == nullptr)
            return 0;
        int val = root->val;
        int right = dfs(root->right, fromParent);
        int left = dfs(root->left, right + val + fromParent);
        root->val = val + right + fromParent;
        return left + val + right;
    }
};